### PR TITLE
Autogenerate Template contextual Terraform Modules

### PR DESCRIPTION
**Pull request by damondouglas**
This PR submits a plugin for autogenerating terraform modules to launch Dataflow Jobs specific to individual templates. **Note to the reviewr: Code under `plugins/*` and `cicd/*` is what mostly needs your review (13 files). The rest is autogenerated terraform or README updates.**

- Updates plugin and related tools [cicd](cicd] and [plugins](plugins) for support in the autogeneration pipeline
- Submits the classic and flex template terraform modules
- Updates the README, using the mvn command, with instructions on where to find and use

Terraform was autogenerated using:
```
mvn clean prepare-package \
  -DskipTests \
  -PtemplatesTerraform
```

READMEs updated using:
```
mvn clean prepare-package \
  -DskipTests \
  -PtemplatesSpec
```

### CODE REVIEW

File: v1/terraform/Stream_GCS_Text_to_BigQuery/dataflow_job.tf [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+		{
+			"name": "color",
+			"type": "STRING"
+		},
+		{
+			"name": "coffee",
+			"type": "STRING"
+		}
+	]
+}"
```
Comment (damondouglas) [inline v1/terraform/Stream_GCS_Text_to_BigQuery/dataflow_job.tf:72]: Use terraform HEREDOC

File: v1/terraform/PubSub_Subscription_to_BigQuery/dataflow_job.tf [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+variable "inputSubscription" {
+  type = string
+  description = "Pub/Sub subscription to read the input from, in the format of 'projects/your-project-id/subscriptions/your-subscription-name'"
+}
+
+variable "outputDeadletterTable" {
+  type = string
+  description = "BigQuery table for failed messages. Messages failed to reach the output table for different reasons (e.g., mismatched schema, malformed json) are written to this table. If it doesn't exist, it will be created during pipeline execution. If not specified, "outputTableSpec_error_records" is used instead."
```
Comment (damondouglas) [inline v1/terraform/PubSub_Subscription_to_BigQuery/dataflow_job.tf:48]: Escape quotes.

File: v1/terraform/GCS_Text_to_Cloud_PubSub/dataflow_job.tf [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+variable "kms_key_name" {
+	type = string
+	description = "The name for the Cloud KMS key for the job. Key format is: projects/PROJECT_ID/locations/LOCATION/keyRings/KEY_RING/cryptoKeys/KEY"
+	default = null
+}
+
+variable "labels" {
+	type = map(string)
```
Comment (damondouglas) [inline v1/terraform/GCS_Text_to_Cloud_PubSub/dataflow_job.tf:76]: Remove extra space

File: plugins/core-plugin/src/main/resources/terraform-classic-template.tf [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+variable "${variable.name}" {
+  type = ${variable.type?lower_case}
+  description = "${variable.description}"
+  <#if variable.defaultValue??>
+  <#if variable.type == "STRING">default = "${variable.defaultValue}"<#else>default = ${variable.defaultValue}</#if>
+  </#if>
+}
+
+</#list>
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-classic-template.tf:46]: Add `google_project_service` block to enable on Dataflow API.

File: plugins/core-plugin/src/main/resources/terraform-flex-template.tf [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+variable "${variable.name}" {
+  type = ${variable.type?lower_case}
+  description = "${variable.description}"
+  <#if variable.defaultValue??>
+  <#if variable.type == "STRING">default = "${variable.defaultValue}"<#else>default = ${variable.defaultValue}</#if>
+  </#if>
+}
+
+</#list>
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-flex-template.tf:46]: Add `google_project_service` block to enable on Dataflow API.

File: plugins/core-plugin/src/main/resources/terraform-flex-template.tf [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+resource "google_dataflow_flex_template_job" "generated" {
+    container_spec_gcs_path = "gs://dataflow-templates-${r"${var.region}"}/latest/flex/${templateName}"
+    parameters = {
+    <#list parameters as variable>
+        ${variable.name} = <#if variable.type == "STRING">var.${variable.name}<#else>tostring(var.${variable.name})</#if>
+    </#list>
+    }
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-flex-template.tf:164]: missing region

File: plugins/core-plugin/src/main/resources/terraform-classic-template.tf [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+resource "google_dataflow_job" "generated" {
+    template_gcs_path = "gs://dataflow-templates-${r"${var.region}"}/latest/${templateName}"
+    parameters = {
+    <#list parameters as variable>
+        ${variable.name} = <#if variable.type == "STRING">var.${variable.name}<#else>tostring(var.${variable.name})</#if>
+    </#list>
+    }
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-classic-template.tf:140]: missing region

File: v1/terraform/Bulk_Compress_GCS_Files/dataflow_job.tf [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+variable "compression" {
+  type = string
+  description = "The compression algorithm used to compress the matched files. Valid algorithms: BZIP2, DEFLATE, GZIP"
+}
+
+variable "outputFilenameSuffix" {
+  type = string
+  description = "Output filename suffix of the files to write. Defaults to .bzip2, .deflate or .gz depending on the compression algorithm."
+  default = ""
```
Comment (damondouglas) [inline v1/terraform/Bulk_Compress_GCS_Files/dataflow_job.tf:59]: set `default = null` instead

File: plugins/core-plugin/src/main/resources/terraform-flex-template.tf [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+	network = var.network
+	num_workers = var.num_workers
+	sdk_container_image = var.sdk_container_image
+	service_account_email = var.service_account_email
+	skip_wait_on_job_termination = var.skip_wait_on_job_termination
+	staging_location = var.staging_location
+	subnetwork = var.subnetwork
+	temp_location = var.temp_location
+}
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-flex-template.tf:184]: Add output block with URL to launched dataflow pipeline job.

File: plugins/core-plugin/src/main/resources/terraform-classic-template.tf [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	max_workers = var.max_workers
+	name = var.name
+	network = var.network
+	service_account_email = var.service_account_email
+	skip_wait_on_job_termination = var.skip_wait_on_job_termination
+	subnetwork = var.subnetwork
+	temp_gcs_location = var.temp_gcs_location
+	zone = var.zone
+}
+
```
Comment (damondouglas) [inline plugins/core-plugin/src/main/resources/terraform-classic-template.tf:156]: Add output block with URL to launched dataflow pipeline job.

**Comment by bvolpato** [comment]
This is so cool! Awesome job @damondouglas!

**Review by liferoad — APPROVED** [review]

**Pull request merged**